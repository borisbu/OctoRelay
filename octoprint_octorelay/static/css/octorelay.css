#settings_plugin_octorelay .tab-content {
    position: relative
}

#settings_plugin_octorelay .btn > input[type='radio'] {
    display: none
}

#settings_plugin_octorelay input.code {
    font-family: monospace;
}

#settings_plugin_octorelay .input-prepend > .add-on.tiny {
    font-size: 0.45rem;
    font-weight: 600;
}

#settings_plugin_octorelay .help-inline a.same-color {
    color: inherit;
}

#settings_plugin_octorelay .control-label span.label,
#settings_plugin_octorelay .help-inline span.label {
    zoom: 0.85; /* not scale */
}

#settings_plugin_octorelay .fa-info {
    border: 2px solid currentColor;
    padding: 3px 8px;
    border-radius: 50%;
    scale: 0.6;
    transform-origin: left;
}

#settings_plugin_octorelay .preview {
    display: flex;
    width: 24px;
    height: 24px;
    overflow: hidden;
    line-height: unset;
    font-size: 1.25rem;
    align-items: center;
    justify-content: center;
    position: absolute;
    top: 0;
    scale: 3;
    transform-origin: top left;
    border: 0.3px dashed #00000066;
    box-sizing: content-box;
    border-radius: 2px;
}

#settings_plugin_octorelay .preview-caption {
    position: absolute;
    top: 75px;
    width: 75px;
    display: flex;
    justify-content: center;
    align-items: baseline;
    gap: 0.5ch;
    white-space: nowrap;
    scale: 0.75;
    transform-origin: top;
}

#navbar_plugin_octorelay > a {
    display: flex;
    float: left;
    width: 40px;
    height: 40px;
    padding: unset;
    cursor: pointer;
    font-size: 1.25rem;
    text-decoration: none;
    align-items: center;
    justify-content: center;
}

#navbar_plugin_octorelay .popover {
    width: auto;
}

#navbar_plugin_octorelay .popover .popover-title {
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 24px;
}

#navbar_plugin_octorelay .popover .popover-content {
    display: flex;
    flex-direction: column;
    gap: 10px;
}

#navbar_plugin_octorelay .popover .popover-content > div {
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: 16px;
}

#navbar_plugin_octorelay .popover-content .btn {
    outline: none;
    margin-top: unset;
}
