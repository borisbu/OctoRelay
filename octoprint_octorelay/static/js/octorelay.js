"use strict";
$(() => {
    const OctoRelayViewModel = function ([settingsViewModel, loginStateViewModel]) {
        const ownCode = "octorelay";
        const self = this;
        self.settingsViewModel = settingsViewModel;
        self.loginState = loginStateViewModel;
        self.onDataUpdaterPluginMessage = function (plugin, data) {
            var _a, _b;
            if (plugin !== ownCode) {
                return;
            }
            const permission = (_b = (_a = self.settingsViewModel.access) === null || _a === void 0 ? void 0 : _a.permissions) === null || _b === void 0 ? void 0 : _b.PLUGIN_OCTORELAY_SWITCH;
            const hasPermission = permission && self.loginState.hasPermission
                ? self.loginState.hasPermission(permission)
                : false;
            const handleClick = (key, value) => {
                const command = () => OctoPrint.simpleApiCommand(ownCode, "update", { pin: key });
                if (!value.confirmOff) {
                    return command();
                }
                const dialog = $("#octorelay-confirmation-dialog");
                dialog.find(".modal-title").text("Turning " + value.labelText + " off");
                dialog
                    .find("#octorelay-confirmation-text")
                    .text("Are you sure you want to turn the " + value.labelText + " off?");
                dialog
                    .find(".btn-cancel")
                    .off("click")
                    .on("click", () => dialog.modal("hide"));
                dialog
                    .find(".btn-confirm")
                    .off("click")
                    .on("click", () => {
                    command();
                    dialog.modal("hide");
                });
                dialog.modal("show");
            };
            for (const [key, value] of Object.entries(data)) {
                const btn = $("#relais" + key).css({
                    display: "flex",
                    float: "left",
                    width: "40px",
                    height: "40px",
                    padding: "unset",
                    cursor: "pointer",
                    "font-size": "1.25rem",
                    "text-decoration": "none",
                    "align-items": "center",
                    "justify-content": "center",
                });
                if (value.active !== undefined) {
                    btn.toggle(hasPermission && value.active === 1);
                }
                if (value.iconText !== undefined) {
                    btn.html(value.iconText);
                }
                if (value.labelText !== undefined) {
                    btn.attr("title", value.labelText);
                }
                btn.off("click").on("click", () => handleClick(key, value));
            }
        };
    };
    OCTOPRINT_VIEWMODELS.push({
        construct: OctoRelayViewModel,
        dependencies: ["settingsViewModel", "loginStateViewModel"],
    });
});
