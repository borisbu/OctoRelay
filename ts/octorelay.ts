interface RelayInfo {
  active: 1 | 0;
  confirmOff: boolean;
  iconText: string;
  labelText: string;
  relay_pin: number;
  state: 1 | 0;
}

type OwnMessage = Record<`r${number}`, RelayInfo>;

type MessageHandler = (plugin: string, data: OwnMessage) => void;

interface OwnProperties {
  settingsViewModel: {
    access?: {
      permissions?: {
        PLUGIN_OCTORELAY_SWITCH?: object;
      };
    };
  };
  loginState: {
    hasPermission?: (permission: object) => boolean;
  };
  onDataUpdaterPluginMessage: MessageHandler;
}

type OwnModel = (
  this: OwnModel & OwnProperties,
  dependencies: object[]
) => void;

$(() => {
  const OctoRelayViewModel: OwnModel = function (
    this,
    [settingsViewModel, loginStateViewModel]
  ) {
    const ownCode = "octorelay";
    const self = this;
    self.settingsViewModel = settingsViewModel;
    self.loginState = loginStateViewModel;

    self.onDataUpdaterPluginMessage = function (plugin, data) {
      if (plugin !== ownCode) {
        return;
      }
      const permission =
        self.settingsViewModel.access?.permissions?.PLUGIN_OCTORELAY_SWITCH;
      const hasPermission =
        permission && self.loginState.hasPermission
          ? self.loginState.hasPermission(permission)
          : false;
      const handleClick = (key: string, value: RelayInfo) => {
        const command = () =>
          OctoPrint.simpleApiCommand(ownCode, "update", { pin: key });
        if (!value.confirmOff) {
          return command();
        }
        const dialog = $("#octorelay-confirmation-dialog");
        dialog.find(".modal-title").text("Turning " + value.labelText + " off");
        dialog
          .find("#octorelay-confirmation-text")
          .text(
            "Are you sure you want to turn the " + value.labelText + " off?"
          );
        dialog
          .find(".btn-cancel")
          .off("click")
          .on("click", () => dialog.modal("hide"));
        dialog
          .find(".btn-confirm")
          .off("click")
          .on("click", () => {
            command();
            dialog.modal("hide");
          });
        dialog.modal("show");
      };
      for (const [key, value] of Object.entries(data)) {
        const btn = $("#relais" + key).css({
          display: "flex",
          float: "left",
          width: "40px",
          height: "40px",
          padding: "unset",
          cursor: "pointer",
          "font-size": "1.25rem",
          "text-decoration": "none",
          "align-items": "center",
          "justify-content": "center",
        });
        if (value.active !== undefined) {
          btn.toggle(hasPermission && value.active === 1);
        }
        if (value.iconText !== undefined) {
          btn.html(value.iconText);
        }
        if (value.labelText !== undefined) {
          btn.attr("title", value.labelText);
        }
        btn.off("click").on("click", () => handleClick(key, value));
      }
    };
  };

  OCTOPRINT_VIEWMODELS.push({
    construct: OctoRelayViewModel,
    dependencies: ["settingsViewModel", "loginStateViewModel"],
  });
});
