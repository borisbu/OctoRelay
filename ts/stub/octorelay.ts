/**
 * WARNING: This is a stub for those who installs the plugin from sources
 *
 * PLEASE NOTE the new URL for installing the distributed package:
 * @link https://github.com/borisbu/OctoRelay/releases/download/latest/release.zip
 *
 * The actual JavaScript is compiled from TypeScript during the execution of release workflow.
 */

$(() => {
  const handleClick = () => {
    const dialog = $("#octorelay-confirmation-dialog");
    dialog.find(".modal-title").text("OctoRelay needs one more update");
    dialog
      .find("#octorelay-confirmation-text")
      .css({
        "font-weight": "normal",
        "font-size": "1rem",
        "word-wrap": "break-word",
      })
      .text(
        "We switched to distributing the plugin in the form of a specially prepared installation " +
          "package instead of offering the installation from sources. " +
          "The URL for fetching updates has changed because of this. " +
          'However, there is no cause for concern: please proceed to the "Software update" section of the ' +
          "OctoPrint settings and update OctoRelay one more time. Sorry for the inconvenience. " +
          "As a last resort, you can always install it manually using the following URL: " +
          "https://github.com/borisbu/OctoRelay/releases/download/latest/release.zip"
      );
    dialog
      .find(".btn-cancel")
      .text("Close")
      .off("click")
      .on("click", () => {
        dialog.modal("hide");
      });
    dialog.find(".btn-confirm").hide();
    dialog.modal("show");
  };
  $("#relaisr1")
    .css({
      display: "flex",
      float: "left",
      width: "40px",
      height: "40px",
      padding: "unset",
      cursor: "pointer",
      "font-size": "1.25rem",
      "text-decoration": "none",
      "align-items": "center",
      "justify-content": "center",
    })
    .html('<i class="fa fa-warning"></i>')
    .attr("title", "OctoRelay: important notice")
    .off("click")
    .on("click", handleClick)
    .show();
});
