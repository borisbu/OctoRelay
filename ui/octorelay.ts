import { initOctoRelayModel } from "./model/initOctoRelayModel";

$(initOctoRelayModel);
