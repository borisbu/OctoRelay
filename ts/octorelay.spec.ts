describe("OctoRelayViewModel", () => {
  const registry: ViewModel[] = [];
  const elementMock: Record<
    | "toggle"
    | "html"
    | "attr"
    | "off"
    | "on"
    | "find"
    | "text"
    | "modal"
    | "css",
    jest.Mock
  > = {
    toggle: jest.fn(),
    html: jest.fn(),
    attr: jest.fn(),
    off: jest.fn(() => elementMock),
    on: jest.fn(),
    find: jest.fn(() => elementMock),
    text: jest.fn(),
    modal: jest.fn(),
    css: jest.fn(() => elementMock),
  };
  const jQueryMock = jest.fn((subject: string | (() => void)) => {
    if (typeof subject === "function") {
      return subject();
    }
    return elementMock;
  });
  const apiMock = jest.fn();
  const permissionsMock: Partial<Record<"PLUGIN_OCTORELAY_SWITCH", object>> = {
    PLUGIN_OCTORELAY_SWITCH: { test: "I am PLUGIN_OCTORELAY_SWITCH" },
  };
  const hasPermissionMock = jest.fn();

  Object.assign(global, {
    OCTOPRINT_VIEWMODELS: registry,
    OctoPrint: { simpleApiCommand: apiMock },
    $: jQueryMock,
  });
  require("./octorelay");

  afterEach(() => {
    jQueryMock.mockClear();
  });

  test("Should push the model into the registry", () => {
    expect(jQueryMock).toHaveBeenCalled();
    expect(registry).toHaveLength(1);
    expect(registry[0]).toMatchSnapshot();
  });

  test("Constructor should set its certain properties", () => {
    const [model] = registry;
    const { construct } = model;
    expect(Object.keys(construct)).toEqual([]);
    construct.call(construct, [
      { access: { permissions: permissionsMock } },
      { hasPermission: hasPermissionMock },
    ]);
    expect({ ...construct }).toMatchSnapshot();
  });

  test("Message handler should ignore other recipients", () => {
    const handler = (registry[0].construct as OwnModel & OwnProperties)
      .onDataUpdaterPluginMessage;
    handler("test", {});
    expect(jQueryMock).not.toHaveBeenCalled();
  });

  test("Message handler should process the supplied configuration", () => {
    hasPermissionMock.mockImplementationOnce(() => true);
    const handler = (registry[0].construct as OwnModel & OwnProperties)
      .onDataUpdaterPluginMessage;
    handler("octorelay", {
      r1: {
        relay_pin: 16,
        inverted_output: false,
        relay_state: true,
        label_text: "Nozzle Light",
        active: true,
        icon_html: "<div>&#128161;</div>",
        confirm_off: false,
      },
      r2: {
        relay_pin: 12,
        inverted_output: true,
        relay_state: true,
        label_text: "Printer",
        active: true,
        icon_html: '<img src="plugin/dashboard/static/img/printer-icon.png">',
        confirm_off: true,
      },
      r3: {
        relay_pin: 18,
        inverted_output: false,
        relay_state: false,
        label_text: "Fan",
        active: true,
        icon_html: '<img src="plugin/dashboard/static/img/fan-icon.png">',
        confirm_off: false,
      },
      r4: {
        relay_pin: 20,
        inverted_output: true,
        relay_state: true,
        label_text: "Chassis Light",
        active: true,
        icon_html: "<div>&#127774;</div>",
        confirm_off: false,
      },
      r5: {
        relay_pin: 24,
        inverted_output: false,
        relay_state: false,
        label_text: "R5",
        active: false,
        icon_html: "ON",
        confirm_off: false,
      },
    });
    expect(hasPermissionMock).toHaveBeenCalledWith({
      test: "I am PLUGIN_OCTORELAY_SWITCH",
    });
    expect(jQueryMock.mock.calls).toMatchSnapshot("$()");
    expect(elementMock.css.mock.calls).toMatchSnapshot(".css()");
    expect(elementMock.toggle.mock.calls).toMatchSnapshot(".toggle()");
    expect(elementMock.html.mock.calls).toMatchSnapshot(".html()");
    expect(elementMock.attr.mock.calls).toMatchSnapshot(".attr()");
    expect(elementMock.off).toHaveBeenCalledTimes(5);
    expect(elementMock.off).toHaveBeenCalledWith("click");
    expect(elementMock.on).toHaveBeenCalledTimes(5);
    expect(elementMock.on).toHaveBeenCalledWith("click", expect.any(Function));

    // clicking on 1st icon, no confirmation
    elementMock.on.mock.calls[0][1]();
    expect(apiMock).toHaveBeenCalledTimes(1);
    expect(apiMock).toHaveBeenCalledWith("octorelay", "update", { pin: "r1" });
    expect(elementMock.on).toHaveBeenCalledTimes(5); // remains

    // clicking on 2nd icon, with confirmation
    elementMock.on.mock.calls[1][1]();
    expect(apiMock).toHaveBeenCalledTimes(1);
    expect(elementMock.find.mock.calls).toMatchSnapshot(".find()");
    expect(elementMock.text.mock.calls).toMatchSnapshot(".text()");
    expect(elementMock.modal).toHaveBeenCalledTimes(1);
    expect(elementMock.modal).toHaveBeenCalledWith("show");
    expect(elementMock.on).toHaveBeenCalledTimes(7);

    // clicking cancel button of the modal
    elementMock.on.mock.calls[5][1]();
    expect(elementMock.modal).toHaveBeenCalledTimes(2);
    expect(elementMock.modal).toHaveBeenLastCalledWith("hide");
    expect(apiMock).toHaveBeenCalledTimes(1); // remains

    // clicking confirm button of the modal
    elementMock.on.mock.calls[6][1]();
    expect(jQueryMock).toHaveBeenCalledWith("#octorelay-confirmation-dialog");
    expect(elementMock.modal).toHaveBeenCalledTimes(3);
    expect(elementMock.modal).toHaveBeenLastCalledWith("hide");
    expect(apiMock).toHaveBeenCalledTimes(2);
    expect(apiMock).toHaveBeenLastCalledWith("octorelay", "update", {
      pin: "r2",
    });
  });

  test.each([
    {
      hasPermission: false,
      permission: permissionsMock.PLUGIN_OCTORELAY_SWITCH,
    },
    {
      hasPermission: true,
      permission: undefined,
    },
  ])(
    "Should not show buttons without permission %#",
    ({ hasPermission, permission }) => {
      hasPermissionMock.mockImplementationOnce(() => hasPermission);
      jest.replaceProperty(
        permissionsMock,
        "PLUGIN_OCTORELAY_SWITCH",
        permission
      );
      const handler = (registry[0].construct as OwnModel & OwnProperties)
        .onDataUpdaterPluginMessage;
      handler("octorelay", {
        r1: {
          relay_pin: 16,
          inverted_output: false,
          relay_state: true,
          label_text: "Nozzle Light",
          active: true,
          icon_html: "<div>&#128161;</div>",
          confirm_off: false,
        },
      });
      expect(hasPermissionMock).toHaveBeenCalledWith({
        test: "I am PLUGIN_OCTORELAY_SWITCH",
      });
      expect(elementMock.toggle).toHaveBeenLastCalledWith(false);
    }
  );
});
