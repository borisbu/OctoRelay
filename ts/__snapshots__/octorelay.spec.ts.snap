// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OctoRelayViewModel Constructor should set its certain properties 1`] = `
{
  "loginState": {
    "hasPermission": [MockFunction],
  },
  "onDataUpdaterPluginMessage": [Function],
  "settingsViewModel": {
    "access": {
      "permissions": {
        "PLUGIN_OCTORELAY_SWITCH": {
          "test": "I am PLUGIN_OCTORELAY_SWITCH",
        },
      },
    },
  },
}
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: $() 1`] = `
[
  [
    "#navbar_plugin_octorelay",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .attr() 1`] = `
[
  [
    "title",
    "Nozzle Light",
  ],
  [
    "title",
    "Printer",
  ],
  [
    "title",
    "Fan",
  ],
  [
    "title",
    "Chassis Light",
  ],
  [
    "title",
    "R5",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .find() 1`] = `
[
  [
    "#relaisr1",
  ],
  [
    "#relaisr2",
  ],
  [
    "#relaisr3",
  ],
  [
    "#relaisr4",
  ],
  [
    "#relaisr5",
  ],
  [
    ".modal-title",
  ],
  [
    "#octorelay-confirmation-text",
  ],
  [
    ".btn-cancel",
  ],
  [
    ".btn-confirm",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .html() 1`] = `
[
  [
    "<div>&#128161;</div>",
  ],
  [
    "<img src="plugin/dashboard/static/img/printer-icon.png">",
  ],
  [
    "<img src="plugin/dashboard/static/img/fan-icon.png">",
  ],
  [
    "<div>&#127774;</div>",
  ],
  [
    "ON",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    "destroy",
  ],
  [
    "destroy",
  ],
  [
    "destroy",
  ],
  [
    "destroy",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .removeAttr() 1`] = `
[
  [
    "data-original-title",
  ],
  [
    "title",
  ],
  [
    "data-original-title",
  ],
  [
    "title",
  ],
  [
    "data-original-title",
  ],
  [
    "title",
  ],
  [
    "data-original-title",
  ],
  [
    "title",
  ],
  [
    "data-original-title",
  ],
  [
    "title",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .removeData() 1`] = `
[
  [
    "original-title",
  ],
  [
    "original-title",
  ],
  [
    "original-title",
  ],
  [
    "original-title",
  ],
  [
    "original-title",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .text() 1`] = `
[
  [
    "Turning Printer off",
  ],
  [
    "Are you sure you want to turn the Printer off?",
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .toggle() 1`] = `
[
  [
    true,
  ],
  [
    true,
  ],
  [
    true,
  ],
  [
    true,
  ],
  [
    false,
  ],
]
`;

exports[`OctoRelayViewModel Message handler should process the supplied configuration: .tooltip() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "placement": "bottom",
    },
  ],
  [
    "destroy",
  ],
  [
    {
      "placement": "bottom",
    },
  ],
  [
    "destroy",
  ],
  [
    {
      "placement": "bottom",
    },
  ],
  [
    "destroy",
  ],
  [
    {
      "placement": "bottom",
    },
  ],
  [
    "destroy",
  ],
  [
    {
      "placement": "bottom",
    },
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 15 and false target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:30:15.000Z" title="8/13/2023, 10:30:15 PM">in 15 seconds</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">OFF</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 15 and true target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:30:15.000Z" title="8/13/2023, 10:30:15 PM">in 15 seconds</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">ON</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 75 and false target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:31:15.000Z" title="8/13/2023, 10:31:15 PM">in 1.3 minutes</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">OFF</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 75 and true target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:31:15.000Z" title="8/13/2023, 10:31:15 PM">in 1.3 minutes</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">ON</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 150 and false target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:32:30.000Z" title="8/13/2023, 10:32:30 PM">in 3 minutes</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">OFF</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 150 and true target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T22:32:30.000Z" title="8/13/2023, 10:32:30 PM">in 3 minutes</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">ON</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 5000 and false target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T23:53:20.000Z" title="8/13/2023, 11:53:20 PM">in 1 hour</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">OFF</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should display upcoming state popover for delay 5000 and true target: .popover() 1`] = `
[
  [
    "destroy",
  ],
  [
    {
      "content": "<time id="time-tag-r1" datetime="2023-08-13T23:53:20.000Z" title="8/13/2023, 11:53:20 PM">in 1 hour</time><button id="cancel-btn-r1" class="btn btn-mini" type="button">Cancel</button>",
      "html": true,
      "placement": "bottom",
      "title": "<span>Nozzle Light goes <span class="label">ON</span></span><button id="pop-closer-r1" type="button" class="close"><span class="fa fa-close fa-sm"></span></button>",
      "trigger": "manual",
    },
  ],
  [
    "show",
  ],
]
`;

exports[`OctoRelayViewModel Should push the model into the registry 1`] = `
{
  "construct": [Function],
  "dependencies": [
    "settingsViewModel",
    "loginStateViewModel",
  ],
}
`;
